### ISSUE TITLE
Narrowing --fixed-cidr doesn't work

### ISSUE THREAD

**Issue by SkyWriter** [docker/docker#17797]
Greetings, ladies and gentlemen!

I've spent a couple of days trying to figure out this issue, but the only thing I've been able to achieve is narrowing it down to the simplest test case:

When I have ```DOCKER_OPTS="--bip=10.2.0.1/16 --fixed-cidr=10.2.0.0/24 --mtu=1400"``` containers can start perfectly fine:

    root@host:~# docker run -it --rm hello-world

    Hello from Docker.
    ...

But having ```DOCKER_OPTS="--bip=10.2.1.1/16 --fixed-cidr=10.2.1.0/24 --mtu=1400"``` breaks things, and trying to say hello to the world yields:

    root@host:~# docker run -it --rm hello-world
    Error response from daemon: Cannot start container 0d337ab1d9145d5e3bb27a242b89f21db4cc3cfb11df545204a55a9d22956f32: [8] 
    System error: failed to set gateway while updating gateway: network is unreachable

To be honest, I am out of my options. My suspicion is that this happens due having the same network addresses in the former case (10.2.0.1/16 == 10.2.0.0/24 bitwise) vs different in the latter (10.2.1.1/16 != 10.2.1.0/24), but I have zero idea of how to resolve it. Since I'm out of options, I'm begging for your help!

I'm using Debian Jessie and Docker version 1.9.0, build 76d6bc9.

Oh, one note on what I'm trying to achieve, just in case I happen to be doing something kinky without even realizing it. What I want is to give a narrower range of IP's to different hosts while keeping a wider to the docker0 interface, then join these docker0's with Open vSwitch/GRE, effectively making all the containers connected across hosts.

Thank you!

**Comment by SkyWriter** [comment]
More tech info as requested:

Docker version:

    root@host:~# docker -v
    Docker version 1.9.0, build 76d6bc9
 
More Docker versions:

    root@host:~# docker version
    Client:
     Version:      1.9.0
     API version:  1.21
     Go version:   go1.4.2
     Git commit:   76d6bc9
     Built:        Tue Nov  3 17:29:38 UTC 2015
     OS/Arch:      linux/amd64

    Server:
     Version:      1.9.0
     API version:  1.21
     Go version:   go1.4.2
     Git commit:   76d6bc9
     Built:        Tue Nov  3 17:29:38 UTC 2015
     OS/Arch:      linux/amd64
    root@host:~# docker info
    Containers: 0
    Images: 8
    Server Version: 1.9.0
    Storage Driver: aufs
     Root Dir: /var/lib/docker/aufs
     Backing Filesystem: extfs
     Dirs: 8
     Dirperm1 Supported: true
    Execution Driver: native-0.2
    Logging Driver: json-file
    Kernel Version: 3.16.0-4-amd64
    Operating System: Debian GNU/Linux 8 (jessie)
    CPUs: 1
    Total Memory: 1000 MiB
    Name: host
    ID: S4QP:JA3I:DL7L:PM4M:MEUA:L2TX:5MIE:CDFE:6D4Z:QCS2:7EOL:3EYU
    WARNING: No memory limit support
    WARNING: No swap limit support

Uname info:

    root@host:~# uname -a
    Linux host 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt11-1+deb8u5 (2015-10-09) x86_64 GNU/Linux

**Comment by cpuguy83** [comment]
Have you seen the `docker network` subcommand?
Docker by default ships with support for vxlan (ie, multihost) networks.
You can also provide a plugin to connect to whatever other backend you want (such as OVS).

**Comment by SkyWriter** [comment]
Yep, I did take a look at it, but (correct me if I'm wrong) it all seemed pretty raw and fragmented at the moment, e.g.:

* http://docs.docker.com/engine/extend/plugins_network/ never actually enumerates network plugins available
* http://docs.docker.com/engine/extend/plugins/ doesn't mention a vxlan plugin (or anything of that kind)

What I found is https://github.com/docker/libnetwork/blob/master/docs/overlay.md, which looked really promising indeed. But overall I have an impression that all of these are pretty bleeding edge features, and it's sensible to wait until the docs settle down.

Or am I just looking in the wrong places?
